### ISSUE TITLE
Attention

### ISSUE THREAD

**Issue by neverix** [xai-org/grok-1#180]
The attention formula has an [unusual line](https://github.com/xai-org/grok-1/blob/d6d9447e2d3c9bd81da571dc8681ea60009e4c03/model.py#L865) that puts attention weights through `tanh`.

1. Is this intentional?
2. What is the reason for this? Normalization/entropy control?
3. Do you have plans to open source an attention kernel that can support this?

**Comment by d9j** [comment]
they just dumped this model  to github and call it a day

**Comment by EwoutH** [comment]
Maybe rewrite the title of this issue to be more descriptive.

**Comment by sebdg** [comment]
I guess it's there to avoid gradient explosion, or vanishing gradients.
Normalizing attention logits using tanh helps in controlling the gradients during the backpropagation process, as it avoids extreme values improving numerical stability.
